### PR TITLE
Feat: V2.2 Support configuration of hash function for source checksums

### PR DESCRIPTION
**Pull request by arnauorriols**
Eszip v2 performs a checksum of its contents. Currently it uses SHA256 to produce the checksum hash. This cryptographic hash function is very computationally expensive. As long as the goal of the hash is to validate the integrity of the source and not it's authenticity, using a cryptographic hash function is unnecessary, and there are alternatives better suited for the job, like CRC32. 

This is becoming an issue now that npm support has bloated eszips considerably, and it is not uncommon to see >100MB eszips for simple projects, specially if they use npm. Here's a comparison of encoding/decoding an eszip using the different checsum options, over the size of the eszip (Apple M1 Pro. Benchmark included in the PR):

image

image

This PR proposes a mechanism to be able to chose the hash function to use for the checksums, and to declare this hash function in the header of the binary encoding.

### CODE REVIEW

File: Cargo.toml
Comment (arnauorriols) [inline Cargo.toml:49]: Do we want to add the different hashes as optional dependencies? FWIW, the current design will degrade to no checksuming (instead of failing the parsing) if the configured hash function is not available

File: src/v2.rs
Comment (arnauorriols) [inline src/v2.rs:711]: Note to reviewer: right not length is known, but the nature of `Options` is thought of as variable length.

File: src/v2.rs
Comment (arnauorriols) [inline src/v2.rs:668]: Question to reviewer: do we want to expose `HashFunction`, or maybe we prefer something like `EszipV2::checksum_with_crc32(&mut self)`, `EszipV2::checksum_with_xxhash(&mut self)`, etc?

File: src/v2.rs
Comment (dsherret) [inline src/v2.rs:1]: Can you provide some background about why we're doing this change and why it includes crc32 and xxhash?

Comment (arnauorriols) [reply src/v2.rs:1]: Added description to the PR

File: src/v2.rs
Comment (arnauorriols) [inline src/v2.rs:567]: @lucacasonato Can you confirm this is no longer needed?

File: src/v2.rs
Comment (dsherret) [reply src/v2.rs:1]: Why do we bother with hashing? At least in the CLI we could do no hash because I don't see a reason why we would do this for deno compile so it would be nice to have an option to not hash at all.

File: src/v2.rs
Comment (dsherret) [inline src/v2.rs:217]: Maybe we should choose one way to do hashing (on or off) then remove the options? (in the CLI it could assume the eszips have no hashing) Obviously backwards compatibility will require sha256 for old eszips.

Comment (arnauorriols) [reply src/v2.rs:217]: We might be able to chose the best hash function for today's context (use cases, hashing state of the art, etc), but we are unable to predict how the context might change in the future. IMO the explicit & flexible approach is worth it in this case.

File: src/v2.rs
Comment (arnauorriols) [reply src/v2.rs:1]: Added support for no checksum, and made it the default for v2.2 and above

File: Cargo.toml
Comment (dsherret) [reply Cargo.toml:53]: I think so. For example, in Deno we have no need for this dep atm

File: Cargo.toml
Comment (dsherret) [inline Cargo.toml:54]: In Deno we use `twox-hash`. This seems to have a BSL license.

File: src/v2.rs
Comment (dsherret) [inline src/v2.rs:231]: Maybe we should pick just one option between Crc32 and XxHash for now? Is there a clear winner between them or does it depend on module size?

**Review by dsherret — COMMENTED** [review]
It looks good to me, but Luca should review this one before landing.

File: src/v2.rs
Comment (dsherret) [inline src/v2.rs:228]: Also, thanks for adding this! It should speed up running code using `deno compile`.

File: src/v2.rs
Comment (arnauorriols) [reply src/v2.rs:231]: I have updated the benchmark to include xxhash and no-checksum. IMO if we make them optional dependencies it should be fine to keep them, but I'm also fine just chosing one.

File: Cargo.toml
Comment (arnauorriols) [reply Cargo.toml:54]: Changed. To my surprise, twox-hash's implementation of xxhash64 is faster than xxhash-rust's xxh3!

Comment (arnauorriols) [reply Cargo.toml]: Added twox-hash as xxhash feature, and kept xxhash-rust as xxhash3 feature. 

I've confirmed twox-hash's xxhash3 implementation is outdated and not producing the same hash as xxhash-rust's. In my machine twox-hash's xxhash64 is faster than xxhash-rust's xxhash3, but different infrastructure might have different results. For this reason, I suggest to keep both.

File: benches/source_hash_function.rs
Comment (arnauorriols) [reply benches/source_hash_function.rs]: They have been removed

File: Cargo.toml
Comment (arnauorriols) [reply Cargo.toml]: I've made them all optional, and kept the "sha256" feature as default for backwards compatibility.

File: deno.json
Comment (arnauorriols) [inline deno.json:3]: @dsherret @lucacasonato FYI I propose to enable all checksums for the wasm build

File: src/v2.rs
Comment (dsherret) [inline src/v2.rs:275]: Perhaps this should error when it finds something it doesn't recognize? Otherwise it might lead to corrupt data being loaded I think.

File: src/v2.rs
Comment (dsherret) [inline src/v2.rs:8]: Is it possible to move this conditional compilation down closer to where they're used within the hash function? Conditional compilation can often be the source of compiler errors when something is misconfigured and putting everything together can help reduce that. For example:

```
Self::XxHash => {
  use std::hash::Hasher;
  let mut hasher = XxHash64::default();
  hasher.write(bytes);
  hasher.finish().to_be_bytes().into()
}
```

File: src/v2.rs
Comment (dsherret) [inline src/v2.rs:239]: Perhaps we should always compile with the enum variants and instead just error when a cargo feature is not enabled when serializing or deserializing with a variant that doesn't have its feature enabled? For example, when reading an eszip that is crc32 we could error saying to enable the `crc32` cargo feature in order to read that eszip.

**Review by dsherret — APPROVED** [review]
LGTM, but I think this is too many hashing functions to support. I think we should have SHA256 and then settle on one faster alternative (probably xxh3).

File: src/v2.rs
Comment (arnauorriols) [reply src/v2.rs:275]: The idea is to have a forward-compatible design that falls back to no checksuming when encountering an unknown hash function. This applies both to decoding an eszip built with a newer version of the library, and built with a hash function not enabled for your library. 

This does not lead to corrupt data thanks to being explicit about the checksum size in the options header. Even if the library does not know how to validate it, it still knows how long the checksum is. Users are expected to use [`EszipV2::is_checksumed()` and `EszipV2::should_be_checksumed()`](https://github.com/denoland/eszip/blob/fa49cb5fa4622b590270976d38855081daaa1f53/src/v2.rs#L744-L762) to ensure the content has been checksumed, if they rely on this guarantee. 

See https://github.com/denoland/eszip/blob/fa49cb5fa4622b590270976d38855081daaa1f53/src/v2.rs#L2444-L2482

File: src/v2.rs
Comment (arnauorriols) [reply src/v2.rs:239]: See https://github.com/denoland/eszip/pull/181#discussion_r1601225605

File: src/v2.rs
Comment (arnauorriols) [reply src/v2.rs]: done

**Pull request merged**
